### PR TITLE
EWPP-373: Ensure the browser widget works also inside IEF.

### PR DESCRIPTION
**Pull request by upchuk**

### CODE REVIEW

File: modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php [hunk 1]
```diff
@@ -282,16 +283,230 @@ public function testFeaturedMediaEntityBrowserWidget(): void {
...
+      ],
+      '1' => [
+        'target_id' => '2',
+        'caption' => 'Image 2 caption',
+      ],
+    ];
+    $actual_values = $node->get('featured_media_field')->getValue();
+    $this->assertEquals($expected_values, $actual_values);
+
+    // Edit the node to remove the current first item in the form (Image 2).
```
Comment (nagyad) [inline modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php:462]: Item 0 is the one you remove and that is Image 1.

File: modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php [hunk 2]
```diff
@@ -282,16 +283,230 @@ public function testFeaturedMediaEntityBrowserWidget(): void {
...
+
+    $remove = $this->getSession()->getPage()->find('css', 'input[data-drupal-selector="edit-article-field-form-inline-entity-form-entities-0-form-featured-media-field-0-current-items-0-remove-button"]');
+    $remove->press();
+    $this->assertSession()->assertWaitOnAjaxRequest();
+
+    // Assert the image was removed from the field.
+    $this->assertSession()->pageTextNotContains('Image 1');
+    $this->getSession()->getPage()->pressButton('Save');
+
+    // Assert the remaining caption should not remain.
```
Comment (nagyad) [inline modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php:477]: This is confusing. You are asserting that the remaining caption can't be saved when the image is removed.

**Review by nagyad — CHANGES_REQUESTED** [review]

**Review by nagyad — DISMISSED** [review]

File: modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php [hunk 3]
```diff
@@ -282,16 +283,230 @@ public function testFeaturedMediaEntityBrowserWidget(): void {
...
+    // Open the IEF form of the embedded article.
+    $this->getSession()->getPage()->pressButton('Edit');
+    $this->assertSession()->assertWaitOnAjaxRequest();
+
+    $remove = $this->getSession()->getPage()->find('css', 'input[data-drupal-selector="edit-article-field-form-inline-entity-form-entities-0-form-featured-media-field-0-current-items-0-remove-button"]');
+    $remove->press();
+    $this->assertSession()->assertWaitOnAjaxRequest();
+
+    // Assert the image was removed from the field.
+    $this->assertSession()->pageTextNotContains('Image 1');
```
Comment (imanoleguskiza) [inline modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php:474]: Here we should double check that the selection button is still available to make sure the original issue does not reappear.

**Review by imanoleguskiza — CHANGES_REQUESTED** [review]

File: modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php [hunk 4]
```diff
@@ -262,6 +276,25 @@ protected function formElementItemsEntities(FieldItemListInterface $items, int $
...
+        $data = reset($data);
+        $value['target_id'] = explode(':', $data)[1];
+        $items->get($delta)->setValue($value);
+        $entity = $entity_storage->load($items[$delta]->target_id);
+        if (!empty($entity)) {
+          $entities[] = $entity;
+        }
+        return $entities;
+      }
+    }
```
Comment (brummbar) [inline modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php:296]: The original widget has a submit callback that sets into the form the value and then rebuild the form. That should be enough to have the values correctly populated. I did delete all this code, copied from the parent widget:
```
'#submit' => [[get_class($this), 'entityBrowserValueUpdated']],
'#limit_validation_errors' => [array_merge($element['#field_parents'], [$this->fieldDefinition->getName()])],
'#executes_submit_callback' => TRUE,
``` 
Then you need to override the method as we handle one value per field only. This way we get completely rid of user input mingling. Tests are green so I assume it's ok.
Also [fago says](https://www.drupal.org/project/entity_browser/issues/2851580#comment-13732789) that the above code seems not useful

**Review by brummbar — CHANGES_REQUESTED** [review]

File: modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php (hunk 3 above)
Comment (upchuk) [reply modules/oe_content_featured_media_field/tests/src/FunctionalJavascript/FeaturedMediaEntityBrowserWidgetTest.php:474]: Indeed. Added assertions. My bad. Not the test fails without the fix.

File: modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php [hunk 5]
```diff
@@ -262,6 +276,25 @@ protected function formElementItemsEntities(FieldItemListInterface $items, int $
...
+    $input_value = NestedArray::getValue($form_state->getUserInput(), $element_path, $input_exists);
+    $value_exists = NestedArray::keyExists($form_state->getValues(), $element_path);
+
+    if (!$value_exists && $input_exists && !is_array($input_value['target_id'])) {
+      $data = empty($input_value['target_id']) ? [] : explode(' ', trim($input_value['target_id']));
+      if ($data) {
+        $data = reset($data);
+        $value['target_id'] = explode(':', $data)[1];
+        $items->get($delta)->setValue($value);
+        $entity = $entity_storage->load($items[$delta]->target_id);
```
[the remark this answers is not in the document]
Comment (upchuk) [reply modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php]: No longer relevant.

File: modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php (hunk 4 above)
Comment (upchuk) [reply modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php]: True. However, it seems I still need to check the input for the case in which an element is removed (and prevent defaults from $items). Could not find a better way. But indeed, using the #submit, we dont need to anymore load values here,

**Review by imanoleguskiza — DISMISSED** [review]

File: modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php (hunk 4 above)
Comment (brummbar) [reply modules/oe_content_featured_media_field/src/Plugin/Field/FieldWidget/FeaturedMediaEntityBrowserWidget.php]: True, it works for the default widget cause they have the same code around.

**Review by brummbar — CHANGES_REQUESTED** [review]
It's all good but we miss test coverage for the removal bug which was fixed with the user input part.

**Review by brummbar — APPROVED** [review]

**Review by imanoleguskiza — APPROVED** [review]

**Pull request merged**